daemon/images.go

```diff
@@ -13,65 +14,112 @@ import (
+	tagPatterns := d.getTagPatterns()
+
+	// Update image for each container
 	for _, service := range services {
 		for _, container := range service.ContainersOrNil() {
+			logger := log.NewContext(logger).With("service", service.ID, "container", container.Name, "currentimage", container.Image)
+
 			currentImageID, err := flux.ParseImageID(container.Image)
 			if err != nil {
-				log.Error(errors.Wrapf(err, "parsing image in service %s container %s (%q)", service.ID, container.Name, container.Image))
+				logger.Log("error", err)
 				continue
 			}
-			if latest := images.LatestImage(currentImageID.Repository()); latest != nil && latest.ID != currentImageID {
-				newImages[latest.ID] = struct{}{}
+
+			repo := currentImageID.Repository()
+			pattern := getTagPattern(tagPatterns, service.ID, container.Name)
+			logger.Log("repo", repo, "pattern", pattern)
+
+			if latest := imageMap.LatestImage(repo, pattern); latest != nil && latest.ID != currentImageID {
+				if err := d.ReleaseImage(service.ID, container.Name, latest.ID); err != nil {
+					logger.Log("error", err, "image", latest.ID)
+				}
 			}
 		}
 	}
+}
+
+func (d *Daemon) unlockedAutomatedServices() ([]flux.ServiceID, error) {
+	automatedServices, err := d.Manifests.ServicesWithPolicy(d.Checkout.ManifestDir(), policy.Automated)
+	if err != nil {
+		return nil, err
+	}
+	lockedServices, err := d.Manifests.ServicesWithPolicy(d.Checkout.ManifestDir(), policy.Locked)
+	if err != nil {
+		return nil, err
+	}
+	return automatedServices.Without(lockedServices).ToSlice(), nil
+}
 
-	// Release the new image(s)
-	for imageID := range newImages {
-		if err := d.NewImage(imageID); err != nil {
-			log.Error(errors.Wrapf(err, "releasing new image %s", imageID))
+func (d *Daemon) getTagPatterns() map[flux.ServiceID]map[string]string {
+	serviceMap := map[flux.ServiceID]map[string]string{}
+	metadata, err := d.Manifests.ServicesMetadata(d.Checkout.ManifestDir())
+	if err != nil {
+		return nil
+	}
+
+	for s, md := range metadata {
+		serviceMap[s] = map[string]string{}
+
+		for k, v := range md {
+			if container, pattern, ok := parseTagPattern(k, v); ok {
+				serviceMap[s][container] = pattern
+			}
 		}
 	}
+
+	return serviceMap
+}
+
+func parseTagPattern(k, v string) (string, string, bool) {
+	if !strings.HasPrefix(k, "flux.weave.works/tag.") || !strings.HasPrefix(v, "glob:") {
+		return "", "", false
+	}
+
+	return strings.TrimPrefix(k, "flux.weave.works/tag."), strings.TrimPrefix(v, "glob:"), true
+}
+
+func getTagPattern(patterns map[flux.ServiceID]map[string]string, serviceID flux.ServiceID, container string) string {
+	if patterns == nil || patterns[serviceID] == nil {
+		return "*"
+	}
+	if p, ok := patterns[serviceID][container]; ok {
+		return p
+	}
+	return "*"
 }
 
-func (d *Daemon) NewImage(imageID flux.ImageID) error {
+func (d *Daemon) ReleaseImage(serviceID flux.ServiceID, container string, imageID flux.ImageID) error {
 	// Try to update any automated services using this image
```

Comment:
This comment is out of date; but it also suggests why things worked in this way -- the idea was to release, in a single go, all the services that used a particular image. This way, the notifications are kept concise.

Now that we are filtering (and specifying particular containers), it's not necessarily the case that a single (new) image will be released to all services that use that image repo. But it would be good to preserve as far as possible the minimal notifications.